Added a faster version of undirected weighted one-mode projection. Wi…

hcars opened the pull request:
To compute the one-mode projection for the bipartite function "weighted_projected_graph", it is faster to multiply the adjacency matrix times its transpose for the undirected case. I changed it to using that for the undirected case of the library function. I show an example of the speed difference in networkx/algorithms/bipartite/tests/test_projection_timing.py. This gets more pronounced for larger networks.

rossbar reviewed (commented):
The test suite isn't the right place for the proposed benchmark and it's causing the tests to fail.

hcars commented:
> The test suite isn't the right place for the proposed benchmark and it's causing the tests to fail.

Hey, Ross, I'll move it real quick

hcars commented:
Ok, I moved the proposed benchmark to the examples/ directory. Would that be the correct location? I, also, refactored the bipartite code for directed graphs to use matrices as well and added a test to the bipartite tests. The test using graph with edges (0, 1), (1, 2), (2, 3) and runs a one-mode projection on {1, 2, 3} to ensure that an edge between 1 and 3 isn't added in the one-mode projection.

hcars commented:
I'm not sure why the 3.8 tests are failing. I my use of numpy imports because that caused a problem, but the scipy imports were already there.

rossbar commented on examples/projection_timing.py:
```diff
@@ -0,0 +1,122 @@
+from networkx.algorithms import bipartite
```
While we certainly appreciate the effort put into demonstrating the performance improvement, NetworkX currently doesn't have a natural place for benchmarks to exist. In practice I think we'd want not to include the benchmark at all.

rossbar commented on networkx/algorithms/bipartite/projection.py:
```diff
@@ -185,33 +185,72 @@ def weighted_projected_graph(B, nodes, ratio=False):
+
+        # Add the edges from the result.
+        denom = 1 / len(top_nodes)
+        for row, col in zip(*projected_adj.nonzero()):
+            val = projected_adj[row, col]
+            if ratio:
+                G.add_edge(nodes[row], nodes[col], weight=val * denom)
             else:
-                weight = len(common) / n_top
-            G.add_edge(u, v, weight=weight)
+                G.add_edge(nodes[row], nodes[col], weight=val)
```
If I'm understanding this correctly, I think you could significantly improve performance (and readability) futher by using array math, e.g. something like

```python
projected_adj = projected_adj.to_array()
if ratio:
    projected_adj = projected_adj / len(top_nodes)
return nx.from_numpy_array(projected_adj)
```

rossbar reviewed (commented):
I'm not familiar enough with projection to comment on the correctness of the implementation (e.g. biadjacency matrix for undirected vs. adjacency matrix for directed. Assuming this is correct though it seems there are some additional opportunities to improve performance further by taking advantage of the arrays instead of looping through them.

dschult commented:
Maybe the right approach is to take out the benchmarking code -- but add a comment to this PR showing the main part of that code and the timing results you obtain. Are we talking about 20% speed-up? or 200% speed-up. :)

hcars commented:
> Maybe the right approach is to take out the benchmarking code -- but add a comment to this PR showing the main part of that code and the timing results you obtain. Are we talking about 20% speed-up? or 200% speed-up. :)

Is there a range of graph sizes v. time that you would like to see?

rossbar commented:
> but add a comment to this PR showing the main part of that code and the timing results you obtain. Are we talking about 20% speed-up? or 200% speed-up. :)

For reference, when I compare implementations using `bipartite.random_graph(2000, 2000, 0.25)` as input, I get:
 - Current implementation in `main`: ~40s
 - Proposed implementation in c72cee7: ~10s
 - [With changes I proposed](https://github.com/networkx/networkx/pull/6410#discussion_r1119619799): ~1.2s

hcars commented:
> > but add a comment to this PR showing the main part of that code and the timing results you obtain. Are we talking about 20% speed-up? or 200% speed-up. :)
> 
> For reference, when I compare implementations using `bipartite.random_graph(2000, 2000, 0.25)` as input, I get:
> 
>     * Current implementation in `main`: ~40s
> 
>     * Proposed implementation in [c72cee7](https://github.com/networkx/networkx/commit/c72cee78ca6668edec8becbd4340d41779f2eba5): ~10s
> 
>     * [With changes I proposed](https://github.com/networkx/networkx/pull/6410#discussion_r1119619799): ~1.2s

Ok, cool, I removed the benchmark file from examples. Here is the result that I got using 10 trials.

Test Network: bipartite.random_graph(2000, 2000, 0.25)

Average Original Method: 89.64801s
Average Proposed Method (Using changes suggested by you): 4.6624s

dschult commented on networkx/algorithms/bipartite/projection.py:
```diff
@@ -185,33 +185,69 @@ def weighted_projected_graph(B, nodes, ratio=False):
+        )
+        G = nx.Graph()
+        G.add_nodes_from(nodes)
+        node_labels = {i: nodes[i] for i in range(len(nodes))}
```
Maybe just `node_labels = dict(enumerate(nodes))`? 

And while we're mapping indexes to nodes, we could map nodes to indexes and use that dict as a lookup tool instead of `<list>.index()` as well as replacing `node in top_nodes` by checking `node not in nodes_indexes`, and replacing `len(top_nodes)`.

Then, another speed-up might involve each place you use `for i in range(len(<list>))`. You should probably use `enumerate` to get both the index and the value so they are both available in that loop.

hcars replied:
Made the change to `node_labels = dict(enumerate(nodes))`. Also, used it for the lookup.

dschult reviewed (commented):
Thanks for this.... I've made a pass through the code. I might have misread some parts, so if my comments are off-base just tell me. See what you think and try anything you think might help to see if it speeds things up.
:)

hcars commented on networkx/algorithms/bipartite/projection.py:
```diff
@@ -185,33 +185,69 @@ def weighted_projected_graph(B, nodes, ratio=False):
+            if type(node) != type(nodes):
+                top_nodes.append(node)
+            elif node not in nodes:
+                top_nodes.append(node)
```
Hey @dschult, I realized this was unnecessary. I just changed it to use list(iterable) to get the list of nodes. The original changes seen here were to help me deal with strings, but I realized the latter changes fixes that for me.

hcars commented on networkx/algorithms/bipartite/projection.py:
```diff
@@ -185,33 +185,69 @@ def weighted_projected_graph(B, nodes, ratio=False):
-                weight = len(common) / n_top
-            G.add_edge(u, v, weight=weight)
+    if not B.is_directed():
+        # Get the biadjacency matrix.
```
I changed the comments here to be more descriptive of what is occuring.

hcars commented on networkx/algorithms/bipartite/projection.py:
```diff
@@ -185,33 +185,69 @@ def weighted_projected_graph(B, nodes, ratio=False):
+            i: all_nodes[i]
+            for i in range(len(all_nodes))
+            if type(all_nodes[i]) == type(nodes) and all_nodes[i] in nodes
+        }
```
Changed to just using dict(enumerate(all_nodes)).

hcars commented on networkx/algorithms/bipartite/projection.py:
```diff
@@ -185,33 +185,69 @@ def weighted_projected_graph(B, nodes, ratio=False):
+
+    else:
+        # Get the adjacency matrix for multiplication.
+        all_nodes = list(B.nodes())
```
Done.
